### ISSUE TITLE
Node count error

### ISSUE THREAD

**Issue by jgillis** [casadi/casadi#1150]
```
from casadi import *                                

states = SX.sym("x",2)                              

ode = SXFunction(daeIn(x=states),[states])          
ode.init()                                          

V = MX.sym("V",2)                                   

g = ode.call(daeIn(x=V))[0]                         

class Bar:                                      
  def __init__(self,p):                         
    self.p = p 

class Foo:                                          
  def __init__(self,a):                             
    self.a = a                                                                                 
    self.bar = Bar(self)                            

f = Foo(g)                                          
```

Gives

```
python: /home/jg/programs/casadi/casadi/core/shared_object.cpp:121: virtual casadi::SharedObjectNode::~SharedObjectNode(): Assertion `count==0' failed.         Aborted    
```

Only manifests itself in Debug mode

**Comment by ghorn** [comment]
I bisected the commits on this issue, compiling in debug mode and running the simple test case above.

The very first commit which exhibits this issue was on March 14, 2014: https://github.com/casadi/casadi/commit/255729000f4f6f292bd01be4bba3ebf1e6aee598

All the commits around that are dealing with references and stuff too: https://github.com/casadi/casadi/commits/master?page=54

The work was all for https://github.com/casadi/casadi/issues/380.
it could still be valid C++ code before and after, just that it interacts with python differently after

**Comment by ghorn** [comment]
so that i remember this, the test case was
```python
from casadi import *
class Foo: pass

f = Foo()
f.a = MX()
f.bar = Foo()
f.bar.p = f
```

and the bisection results:
```
git bisect start
# bad: [e78c29a5de9aa81ac7445658565883a1448cb9e0] Merging release-2.2.0 to master
git bisect bad e78c29a5de9aa81ac7445658565883a1448cb9e0
# bad: [8b57580d9bed7f74c3547daaa83cb208d405aca5] Merge branch 'develop' of github.com:casadi/casadi into develop
git bisect bad 8b57580d9bed7f74c3547daaa83cb208d405aca5
# bad: [00cbc0b6dfb4556968cba4924398fa7c6335e2f0] json: fix baselist
git bisect bad 00cbc0b6dfb4556968cba4924398fa7c6335e2f0
# good: [efcaff32080dc0d8d3b908847c7340bd2e3f7d5a] Revert "Commented out failing unit tests, #884"
git bisect good efcaff32080dc0d8d3b908847c7340bd2e3f7d5a
# good: [efcaff32080dc0d8d3b908847c7340bd2e3f7d5a] Revert "Commented out failing unit tests, #884"
git bisect good efcaff32080dc0d8d3b908847c7340bd2e3f7d5a
# good: [f7cf2543a2f76d276e25419da0655e64874781f9] Refactored fx.hpp, #994 #1020
git bisect good f7cf2543a2f76d276e25419da0655e64874781f9
# good: [f7cf2543a2f76d276e25419da0655e64874781f9] Refactored fx.hpp, #994 #1020
git bisect good f7cf2543a2f76d276e25419da0655e64874781f9
# bad: [12834774da5cd67fda576b375cef9bcbabbb0a32] #1041 Added test script
git bisect bad 12834774da5cd67fda576b375cef9bcbabbb0a32
# bad: [f641efd2bb97fda10464ace85bf8de8a73977d7c] fixup! Fixes #1022: IPOPT segfault for trivial problem
git bisect bad f641efd2bb97fda10464ace85bf8de8a73977d7c
# bad: [f641efd2bb97fda10464ace85bf8de8a73977d7c] fixup! Fixes #1022: IPOPT segfault for trivial problem
git bisect bad f641efd2bb97fda10464ace85bf8de8a73977d7c
# bad: [780bd04d02dd7380673faf8620b557c2752b43de] delete trailing whitespace in top-level CMakeLists
git bisect bad 780bd04d02dd7380673faf8620b557c2752b43de
# bad: [780bd04d02dd7380673faf8620b557c2752b43de] delete trailing whitespace in top-level CMakeLists
git bisect bad 780bd04d02dd7380673faf8620b557c2752b43de
# good: [b609b829fdee571cdf47297fbf2e07feaea775df] MX::isNull --> MX::isNullE in CallFX, #380
git bisect good b609b829fdee571cdf47297fbf2e07feaea775df
# bad: [ed6b65d7a61834b264878f686311c8d9597cdce3] Better sparsity pattern calculation #1021
git bisect bad ed6b65d7a61834b264878f686311c8d9597cdce3
# bad: [ca6e0c775e85d66fc429c2aa39cbc59613ab00df] Simplifications in MX, post #380
git bisect bad ca6e0c775e85d66fc429c2aa39cbc59613ab00df
# bad: [255729000f4f6f292bd01be4bba3ebf1e6aee598] Changed default constructor of MX to return a 0-by-0 matrix instead of a null pointer, #380
git bisect bad 255729000f4f6f292bd01be4bba3ebf1e6aee598
# good: [ebcc8e7824854960205337c35016a2b376629274] MX::isNull --> MX::isNullE in FXInternal, #380
git bisect good ebcc8e7824854960205337c35016a2b376629274
# good: [83a5b2969a02328517be18a64e87d8bd031f5c11] MX::isNull --> MX::isNullE in XFunctionInternal, #380
git bisect good 83a5b2969a02328517be18a64e87d8bd031f5c11
# first bad commit: [255729000f4f6f292bd01be4bba3ebf1e6aee598] Changed default constructor of MX to return a 0-by-0 matrix instead of a null pointer, #380
```

**Comment by ghorn** [comment]
After much investigation, I know why this happens for `MX()` but not `MX.sym('x')` or `SX` or `Function`. For all of these objects, there is a memory leak from the circular references because destructors can't be called. The difference is that `MX()` returns a singleton `MXNode` representing `0x0`, where the `0x0` node is statically allocated. When `exit()` is called, all statically allocated objects are destroyed, including this singleton who's MX was leaked. This can be verified by calling `quick_exit()` and seeing that there is no reference count failure.
